### PR TITLE
Boost warnings

### PR DESCRIPTION
**Pull request by bennibbelink**
The Boost deprecation warnings are gone now.  

The most significant changes involve converting `<boost/detail/sp_typeinfo.hpp>` to `<boost/core/typeinfo.hpp` and its declarations.

I also cleaned up some unnecessary `#include`s and added some `#include`s to explicitly declare `<boost/shared_ptr.hpp>` where it had been declared implicitly through other header files (not necessary but might prevent issues in the future)

### CODE REVIEW

**Review by gonuke — APPROVED** [review]
This looks great!! I thought it might be straightforward, but they did make a clutter. Thanks @bennibbelink

**Comment by gonuke** [comment]
I'll merge once this goes out of draft

**Comment by bennibbelink** [comment]
Thanks! It was more straightforward than I was expecting thankfully.  I'll move this out of draft once we get #1608 merged, I haven't tested these boost changes on 20.04 so I'd like to see the new testing matrix run first

**Review by gonuke — APPROVED** [review]

**Pull request merged**
